Update CMakeLists.txt

unmovingcastle opened the pull request:
A PR to go with [pueoBuilder's CleanUp branch](https://github.com/PUEOCollaboration/pueoBuilder/pull/16)
and [libAntarcticaRoot's updateCMake branch](https://github.com/PUEOCollaboration/libAntarcticaRoot/pull/2)
and [NiceMC's updateCMake branch](https://github.com/PUEOCollaboration/nicemc/pull/45)
and [pueoSim's updateCMake branch](https://github.com/PUEOCollaboration/pueoSim/pull/70)
and [pueo-data's updateCMake branch](https://github.com/PUEOCollaboration/pueo-data/pull/2)
and [pueoEvent's updateCMake branch](https://github.com/PUEOCollaboration/pueoEvent/pull/2)

**THESE PRs NEED TO BE MERGED SIMULTANEOUSLY**

## general
1. Default option is to use the **old** `CMakeLists.txt` (`cmake/OldCMakeLists.cmake)` for backward compatibility; can turn on the option to use the **new** `CMakeLists.txt` (ie. this pull request) with `DNEW_CMAKELISTS=ON`.
1. Using `target_*()` commands whenever possible to avoid adding things globally.
1. Packaged the project for downstream repositories to `find_package()` and obtain the `PUBLIC` compiler flags, preprocessor macros, header files, and libraries automatically when they link against RootFftwWrapper; resources:
[modern way to package things](https://blog.vito.nyc/posts/cmake-pkg/)
1. Now using `find_package(CONFIG)` mode and linking through `Namespace::target` wherever possible; e.g. [modern way to link against OpenMP](https://stackoverflow.com/questions/12399422/how-to-set-linker-flags-for-openmp-in-cmakes-try-compile-function/51448364#51448364)
1. Separated build tree and install tree include paths so that downstream `find_package()` can correctly identify `RootFftwWrapper's` include path automatically; `RootFftwWrapper`'s header files now grouped together in a single directory in the install tree
1. Linking against fftw libraries `PUBLIC`ally (**most likely unnecessary because no downstream project should explicitly need fftw and if anything they should simply use the `FindFFTW` module themselves**).

## compiler flags and definitions

1. `-Os` removed: `O2` already in RelWithDebugInfo and other build types
2. `-g` removed: already in the build types Debug and RelWithDebugInfo, etc.
1. Only use warnings flags `-Wall` and `-Wextra` in `Debug` mode; resources:
[Modern CMake generator expression](https://cliutils.gitlab.io/modern-cmake/chapters/basics/functions.html#generator-expressions)
1. Checked the various options' associated compiler flags and preprocessor definitions; made sure these flags are `PRIVATE` unless they are needed downstream.
1. Only install VCL headers if option VECTORIZE is on; made sure to propagate the preprocessor macro `ENABLE_VECTORIZE` for `pueoSim`.

## CERN ROOT: dictionary and linking
1. Checked all header files and determined the `ROOT::<targets>` to link against  (instead of linking against the entire ROOT library `${ROOT_LIBRARIES}`). See CERN ROOT's [CMake project example](https://root.cern/manual/integrate_root_into_my_cmake_project/#full-example-event-project). The checklist is stored in `cmake/rootLibList.txt`.
1. Automated CERN ROOT dictionary generation to reduced a few lines; resources:
[CERN ROOT Forum](https://root-forum.cern.ch/t/usage-of-include-directories-with-root-generate-dictionary/45378/2?u=unmovingcastle)
[Modern CMake CERN ROOT section](https://cliutils.gitlab.io/modern-cmake/chapters/packages/ROOT.html#building-the-dictionary-with-an-existing-target)
1. Linking against `ROOT::<targets>` `PRIVATE`ly to avoid redundant linking downstream;

## CERN ROOT: flags
In conclusion, 
1.  No longer using `$ROOT_CXX_FLAGS`
1. `pipe` and `fsigned-char` within `$ROOT_CXX_FLAGS` manually set by `target_compile_options(...)`
1. `pthread` [taken care of by linking](https://stackoverflow.com/a/29871891/21955752)
1. `cxx_std_*` is [set automatically](https://root-forum.cern.ch/t/getting-the-c-standard-if-root-cxx-standard-is-not-set/59439) from `${ROOT_CXX_FLAGS}` **or** `${ROOT_CXX_STANDARD}`; once set in `RootFftwWrapper` as `PUBLIC`, the standard [propagates downstream](https://cliutils.gitlab.io/modern-cmake/chapters/features/cpp11.html).

Below are some trial-and-error results that motivated the above change.
* The CMake variable `$ROOT_CXX_FLAGS` differs between platforms, on macOS (platform: M1 macOS with Docker Fedora 41) these are 
```-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -pipe -fsigned-char -pthread```. Simply using all of `$ROOT_CXX_FLAGS` leads to run-time error in pueoSim on macOS, but 
completely getting rid of `$ROOT_CXX_FLAGS` also leads to error because `-fsigned-char` is needed.
* Hence, the flags should probably be manually handled instead. I'm assuming Midway3 is our standard computing environment, on which `$ROOT_CXX_FLAGS` contain: ```-std=c++17, -pthread, -fsigned-char and -pipe```.
*  ROOT claims to propagate this information if we use the `ROOT::target` syntax, but `cxx_std_17` (or whatever standard that ROOT is compiled with) is still explicitly set to be safe (`cmake/modules/getRootStandard.cmake`). With **ROOT 6.30.06+** this is simply obtained through `${ROOT_CXX_STANDARD}` but this variable is not populated on Midway or OSC; fortunately it could also be obtained through `${ROOT_CXX_FLAGS}` (although not on mac).
* `pthread` appears to be unnecessary but is kept

Commit:
initial commit

Commit:
using project version automatically

Commit:
using target include instead of global include

Commit:
using project name automatically

Commit:
doing away with compatlib, for now

Commit:
doing away with environment variables.

Commit:
using the ROOT finder shipped by ROOT

Commit:
removing legacy code

Commit:
target include dir for binary

Commit:
pueoBuilder already checks for Minuit2

hopefully we all have this now

Commit:
using CMAKE_CURRENT_SOURCE_DIR instead

instead of CMAKE_SOURCE_DIR

Commit:
Generating ROOT dict automatically according to:

https://root-forum.cern.ch/t/usage-of-include-directories-with-root-generate-dictionary/45378/2?u=unmovingcastle
https://cliutils.gitlab.io/modern-cmake/chapters/packages/ROOT.html#building-the-dictionary-with-an-existing-target

Commit:
separatig build and install trees

I don't know why exactly the generator expression is needed
or how it affects the build and install trees, but this seems
pretty standard, see:

https://twdev.blog/2024/02/cmake_external_project/
and
https://cliutils.gitlab.io/modern-cmake/chapters/basics/functions.html#generator-expressions

Commit:
More boiler plate code
which I don't fully understand, but works

Commit:
done for now, haven't looked into the options

Commit:
lib64 -> lib

ie not using ${CMAKE_INSTALL_LIBDIR}

also, fixed a bug (...Targets.cmake)

Commit:
minor tweaks before modernizing the options

Commit:
using target_compile_options instead of set(...)

Commit:
moving option handling all the way up

Commit:
done updating linalg and vector options

Commit:
openMP option checked, not working currently

Commit:
checked force_OLD_ABI; got an error

something about linking with ROOT::Minuit2,
but this is probably because I shouldn't be using this option?

Commit:
done with all options

Commit:
removing namespace requirement

Commit:
avoided hardcoding eigen3 and armadillo options

Commit:
target_compile_options -> target_compile_definitions

Commit:
openmp fixed

modern way to use openmp with cmake
(https://stackoverflow.com/a/51448364/21955752)

Commit:
PRIVATE->PUBLIC

Commit:
cleaner install paths

Commit:
use ROOT::<targets> instead of ${ROOT_LIBRARIES}

as specified by the documentation:
https://root.cern/manual/integrate_root_into_my_cmake_project/#full-example-event-project

TODO: still not sure which ones we use though...

Commit:
properly linked against all the ROOT::libraries

Commit:
added a housekeeping txt file

to track the CERN ROOT header files are used and their corresponding cmake targets (ie. libraries)

Commit:
double-checked the list, going to bed now

Commit:
ROOT::FitPanel to replace ROOT::Physics

Commit:
only use debug flags in debug mode

Commit:
dependencies (ie FFTW) handled

Commit:
removing the fftw3 dependency check (for now)

Commit:
simplified/modernized a bit

no need to write @package init@ in the config file
(actually, no need to "generate" the config file)
see:
https://blog.vito.nyc/posts/cmake-pkg/

Commit:
only use/install vectorclass if VECTORIZE

Commit:
bug fix for the previous commit

Commit:
using -march=native if NATIVE_ARCH is on

previously this would require VECTORIZE=ON as well, this behavior is
removed

Commit:
Added ROOT_CXX_FLAGS back, propagates downstream

previoulsy accidentally removed.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
-do_binary(testFFTtools) 
+# TODO: FFTW didn't ship with a finder?
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules/")
+find_package(FFTW REQUIRED)
```
On most systems it probably ships with a .pc file, so you can probably use PkgConfig stuff to find it?  But the cmake module allows overwriting the directory by env var at least

unmovingcastle replied:
Okay. Let's just stick with this finder for now. I suppose it's not terrible to have to use MODULE mode just this once.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
-if( ${ROOT_VERSION} VERSION_GREATER "5.99.99")
-    add_custom_target(${DICTNAME}.pcm DEPENDS ${DICTNAME})
+cmake_minimum_required(VERSION 3.19)
+project(RootFftwWrapper VERSION 3.0.1)
```
probably need to bump up the version number since there are breaking changes?

unmovingcastle replied:
There are no actual change to the source code though, and there will be *no* breaking change at all, unless the user specifically opt-in with `-DNEW_CMAKELISTS=ON` when building `RootFftwWrapper`; even then, there's still no change in what's being compiled & installed.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
+project(RootFftwWrapper VERSION 3.0.1)
+# set(compatlibname "RootFftwWrapperCompat")
+
+file(GLOB SOURCE_FILES "${CMAKE_CURRENT_SOURCE_DIR}/src/*.cxx")
```
if we're going to fix up the CMakeLists, we should probably remove the globbing, since globbing is failure prone. (see e.g. note in https://cmake.org/cmake/help/latest/command/file.html#glob)

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
+file(GLOB SOURCE_FILES "${CMAKE_CURRENT_SOURCE_DIR}/src/*.cxx")
+# set (COMPAT_SOURCE_FILES src/FFTtools.cxx src/FFTWComplex.cxx src/RFSignal.cxx src/RFFilter.cxx ) 
+add_library(${PROJECT_NAME} SHARED ${SOURCE_FILES})
+# add_library(${compatlibname} SHARED ${COMPAT_SOURCE_FILES}) 
```
why get rid of the compat lib? There was at least one user...

unmovingcastle replied:
Since there's an option to use the original `CMakeLists.txt` and PUEO is probably the only one using the new `CMakeLists.txt`, I'd say we don't build the compat version. Unless one of us is using it?

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
-    set(LD_UTIL $ENV{ANITA_UTIL_INSTALL_DIR}/lib)
-    set(INC_UTIL $ENV{ANITA_UTIL_INSTALL_DIR}/include)
-    do_install()
+if(NOT WIN32)
```
shouldn't this depend on TERM and not platform?

unmovingcastle replied:
I'll just go ahead and remove all the colors. They aren't really being used here in `libRootFftwWrapper`.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
-# Install in any and all available locations.
-# This might lead to some unnecessary re-installation (if you're running both ANITA and PUEO software, e.g.
-# To avoid this, disable the irrelevant environmental variable when building either
-if(DEFINED ENV{PUEO_UTIL_INSTALL_DIR})
```
getting rid of these env variables will break libRootFftwWrapper for almost all users.  Probably this does need to get cleaned up into a better macro though.

unmovingcastle replied:
The install location will be identical to the original install location for PUEO users, since the environment variable `PUEO_UTIL_INSTALL_DIR` is still kept and points to the install location of the PUEO software suite (see superbuild `pueoBuilder`'s `CMakeLists.txt`)

For all other users, they will default to using the original `CMakeLists.txt` of `RootFftwWrapper` so there will be no breaking changes there either.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
+#                                       BUILDING
+#================================================================================================
 
+file(GLOB HEADER_FILES "${CMAKE_CURRENT_SOURCE_DIR}/include/*.h")
```
ditto on GLOB

cozzyd commented on cmake/RootFftwWrapperConfig.cmake:
```diff
@@ -0,0 +1,7 @@
+include("${CMAKE_CURRENT_LIST_DIR}/RootFftwWrapperTargets.cmake")
```
is this file supposed to exist?

unmovingcastle replied:
This is generated by CMake automatically so that downstream projects can use `find_package()` to find `RootFftwWrapper` as a `target`, or something.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
+endif()
 
+# only use warning flags in debug mode
+target_compile_options(${PROJECT_NAME} PRIVATE $<$<CONFIG:Debug>:-Wall -Wextra>)
```
I don't think this is a good idea?

unmovingcastle replied:
Okay, changed warnings to opt-out.
The idea is that there should be a choice to turn these off.
Now the behavior is this: 
1. default build mode is `RelWithDebInfo` which also gives us `-g -DNDEBUG -O2`
2. attach warning flags if `RelWithDebInfo`; users can automatically have these turned off if they want to use the `Release` mode instead.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
+#================================================================================================
+#                                       OPTIONS
+#================================================================================================
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NATIVE_ARCH AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 4.1) 
```
why get rid of the NATIVE_ARCH?  It is useful on heterogeneous clusters.

unmovingcastle replied:
I don't know what heterogeneous clusters are.
Anyways, `NATIVE_ARCH` was never an option in the old `CMakeLists.txt` either.
This is just rewriting the logic so that we don't have nested `if`s.

cozzyd commented on CMakeLists.txt:
```diff
@@ -1,375 +1,222 @@
-
-if (NOT CMAKE_BUILD_TYPE) 
-  set (CMAKE_BUILD_TYPE Default
-    CACHE STRING "Choose tye type of build: None Debug Release RelWithDebInfo MinSizeRel Default"
```
do you know what Default entails if we don't specify it? Is it platform dependent? I can't quickly work out what it does. On my system (Fedora) it seems to be roughly equivalent to RelWithDebInfo (based on doing make VERBOSE=1) ?   But at least in the past on some systems it was not that...

unmovingcastle replied:
On my computer (M1 macOS running Docker Fedora), using the CMake gui:
`cd build; ccmake .`
and then toggle with `t`, I see that DEFAULT basically just doesn't add any flags.

unmovingcastle replied:
Actually it's probably better to default to using `RelWithDebInfo` which would automatically add `-g -O2 -DNDEBUG`

Commit:
commented out options originally commented out

Commit:
tripple checked ROOT::<targets>

Commit:
installing fftw header files for downstream

ie. copying the header files fftw3* and installing them under
<install_prefix>/RootFftwWrapper/fftw/ so that downstream projects
can find these headers files when they link against RootFftwWrapper,
WITHOUT needing the FindFFTW.cmake module.

Commit:
still installing the FindFFTW.cmake module

Commit:
installing ROOT pcm files so it doesn't yell at me

Commit:
TODO: can't have ROOT_CXX_FLAGS ???

These flags were originally intended to be used (see Master branch)
but now if I use them in RootFftwWrapper, peuoSim seg faults.

Also, if I put these things in AntarcticaRoot, everything seems fine.

Commit:
dealt with ROOT_CXX_FLAGS

Commit:
keeping Threads, throwing away standards

Commit:
c++ standard, getting more verbose now...

Commit:
gnu++17 no; c++17 yes

Commit:
no longer installing fftw header files

Commit:
fixed a missing )

Commit:
un-breaking the cmakelist for all other collaborations

Commit:
forgot the commit this, no actual change

Commit:
no change

cozzyd commented on cmake/NewCMakeLists.cmake:
```diff
@@ -0,0 +1,243 @@
+elseif(FFTTOOLS_ENABLE_THREAD_SAFE)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE FFTTOOLS_THREAD_SAFE)
+
+elseif(FFTTOOLS_ENABLE_OPENMP)
```
this is a very impotant option, it should probably be exposed to pueoBuilder?

cozzyd commented on cmake/NewCMakeLists.cmake:
```diff
@@ -0,0 +1,243 @@
+# note that the flag -std=cxx_std_* may be omitted by CMake if requested standard < default standard
+# see (https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support)
+
+# pretty sure pthread is not actually needed, but I guess it doesn't hurt to keep this
```
it is if built with one of the parallelization options

cozzyd reviewed (commented):
are we giving up on upstreaming this?

unmovingcastle commented:
> are we giving up on upstreaming this?

No, we are not giving up on upstreaming this. I believe you are looking at an outdated version of the CMakeLists.txt. About three weeks ago there was a commit that added an option that allows users to switch between the original vs. new version (default is to use the original CMakeLists.txt) and `pueoBuilder` would opt in to use the new version.

unmovingcastle commented:
## TODOs
- [x] get rid of `FORCE_OLD_ABI` in the `NewCMakeLists.txt`.
- [x] remove globbing.
- [x] expose `OpenMP` to `pueoBuilder`
- [x] warning flags by default (TBD)
- [ ] ~put `NATIVE_ARCH` back (TBD)~

Commit:
got rid of colors (not used)

Commit:
got rid of FORCE_OLD_ABI

Commit:
OpenMP: PUBLIC (so test scripts can link against)

Commit:
screw it, PUBLIC->PRIVATE

the test script can link against OpenMP themselves.

Commit:
no more globbing

Commit:
Use RelWithDebInfo and make warning flags opt-out

Commit:
trying to use CMake feature FILE_SET

Commit:
no need to explicitly include fftw headers ourselves

Commit:
cleaning up install commands

Commit:
requiring cmake 3.23
